File: javacommons/javastorage/tsrc/java_api/javasrc/com/nokia/mj/test/storage/TestUpdate.java
/*
* Copyright (c) 2008-2009 Nokia Corporation and/or its subsidiary(-ies).
* All rights reserved.
* This component and the accompanying materials are made available
* under the terms of "Eclipse Public License v1.0"
* which accompanies this distribution, and is available
* at the URL "http://www.eclipse.org/legal/epl-v10.html".
*
* Initial Contributors:
* Nokia Corporation - initial contribution.
*
* Contributors:
*
* Description:
*
*/


package com.nokia.mj.test.storage;

import com.nokia.mj.test.storage.utils.StorageSessionTestUtils;
import com.nokia.mj.impl.storage.*;

import com.nokia.mj.impl.installer.utils.InstallerMain;
import j2meunit.framework.Test;
import j2meunit.framework.TestCase;
import j2meunit.framework.TestMethod;
import j2meunit.framework.TestSuite;

/**
 * StorageSession write test cases. See test methods for test case details.
 */
public class TestUpdate extends TestCase implements InstallerMain, StorageNames
{
    /**
     * Directory for JavaStorage tests.
     */
    private static final String iTestRoot = "./jstest";

    /**
     * Directory for JavaStorage journal and temp files.
     */
    private static final String iIsRoot = iTestRoot + "/js";

    private StorageSession iSession = null;
    private StorageSessionTestUtils iJtu = null;

    public void installerMain(String[] args)
    {
        TestSuite suite = new TestSuite(this.getClass().getName());

        suite.addTest(new TestUpdate("testUpdateArguments", new TestMethod()
        {
            public void run(TestCase tc)
            {
                ((TestUpdate)tc).testUpdateArguments();
            }
        }));

        suite.addTest(new TestUpdate("testUpdate", new TestMethod()
        {
            public void run(TestCase tc)
            {
                ((TestUpdate)tc).testUpdate();
            }
        }));

        com.nokia.mj.impl.utils.OmjTestRunner.run(suite);
    }

    public TestUpdate()
    {
    }

    public TestUpdate(String aTestName, TestMethod aTestMethod)
    {
        super(aTestName, aTestMethod);
    }

    protected void setUp()
    {
        iSession = StorageFactory.createSession();
        iJtu = new StorageSessionTestUtils();
    }

    protected void tearDown()
    {
        if (iSession != null)
        {
            try
            {
                iSession.destroySession();
            }
            catch (StorageException se)
            {
                // No can do
                System.out.println("TearDown failed: " + se.toString());
            }
        }
    }

    /**
     * Test update arguments.
     * 1. Test empty tableName throws SE.
     * 2. Test null tableName throws SE.
     * 3. Test empty update entry throws SE.
     * 4. Test null update entry throws SE.
     * 5. Test non-existent table throws SE.
     * 6. Test update non-existent attribute throws SE.
     * 7. Test update no connection throws SE.
     */
    public void testUpdateArguments()
    {
        String tableName = null;
        StorageEntry updateEntry = null;
        StorageEntry matchEntry = null;
        String appUid = "UpdateMe";

        try
        {
            iSession.open();
            iSession.startTransaction();

            assertTrue("PreCondition failed",
                       iJtu.populateWithDefault(iSession, appUid));

            assertTrue("PreCondition: population failed",
                       iJtu.checkDefaultParams(iSession, appUid));
        }
        catch (StorageException se)
        {
            // PASSSED
        }
        catch (Throwable t)
        {
            assertTrue("UnExpected exp0: " + t.toString(), false);
        }

        // 1. Test empty tableName throws SE.
        try
        {
            iSession.open();
            tableName = "";

            updateEntry = new StorageEntry();
            updateEntry.addAttribute(
                new StorageAttribute(VENDOR,
                                     "TestUpdateVal"));

            matchEntry = new StorageEntry();
            matchEntry.addAttribute(
                new StorageAttribute(ID, appUid));

            iSession.update(tableName, updateEntry, matchEntry);
            assertTrue("No exp. when empty table name", false);
        }
        catch (StorageException se)
        {
            // PASSED
        }
        catch (Throwable t)
        {
            assertTrue("UnExpected exp1: " + t.toString(), false);
        }

        iSession.close();

        // 2. Test null tableName throws SE.
        try
        {
            iSession.open();
            tableName = null;
            appUid = "UpdateMe";

            updateEntry = new StorageEntry();
            updateEntry.addAttribute(
                new StorageAttribute(VENDOR,
                                     "TestUpdateVal"));

            matchEntry = new StorageEntry();
            matchEntry.addAttribute(
                new StorageAttribute(ID, appUid));

            iSession.update(tableName, updateEntry, matchEntry);
            assertTrue("No exp. when null table name", false);
        }
        catch (StorageException se)
        {
            // PASSED
        }
        catch (Throwable t)
        {
            assertTrue("UnExpected exp2: " + t.toString(), false);
        }

        iSession.close();

        // 3. Test empty update entry throws SE.
        try
        {
            iSession.open();
            tableName = APPLICATION_PACKAGE_TABLE;

            updateEntry = new StorageEntry();

            matchEntry = new StorageEntry();
            matchEntry.addAttribute(
                new StorageAttribute(ID, appUid));

            iSession.update(tableName, updateEntry, matchEntry);
            assertTrue("No exp. when empty entry", false);
        }
        catch (StorageException se)
        {
            // PASSED
        }
        catch (Throwable t)
        {
            assertTrue("UnExpected exp3: " + t.toString(), false);
        }

        iSession.close();

        // 4. Test null update entry throws SE.
        try
        {
            iSession.open();
            tableName = APPLICATION_PACKAGE_TABLE;
            updateEntry = null;

            matchEntry = new StorageEntry();
            matchEntry.addAttribute(
                new StorageAttribute(ID, appUid));

            iSession.update(tableName, updateEntry, matchEntry);
            assertTrue("No exp. when null entry", false);
        }
        catch (StorageException se)
        {
            // PASSED
        }
        catch (Throwable t)
        {
            assertTrue("UnExpected exp4: " + t.toString(), false);
        }

        iSession.close();

        // 5. Test non-existent table throws SE.
        try
        {
            iSession.open();
            tableName = "FOO_BAR";
            appUid = "UpdateMe";

            updateEntry = new StorageEntry();
            updateEntry.addAttribute(
                new StorageAttribute(VENDOR,
                                     "TestUpdateVal"));

            matchEntry = new StorageEntry();
            matchEntry.addAttribute(
                new StorageAttribute(ID, appUid));

            iSession.update(tableName, updateEntry, matchEntry);
            assertTrue("No exp. when non-existent table", false);
        }
        catch (StorageException se)
        {
            // PASSED
        }
        catch (Throwable t)
        {
            assertTrue("UnExpected exp5: " + t.toString(), false);
        }

        iSession.close();

        // 6. Test update non-existent attribute throws SE.
        try
        {
            iSession.open();
            tableName = APPLICATION_PACKAGE_TABLE;
            appUid = "UpdateMe";

            updateEntry = new StorageEntry();
            updateEntry.addAttribute(
                new StorageAttribute(VENDOR,
                                     "TestUpdateVal"));

            updateEntry.addAttribute(
                new StorageAttribute("FOO_BAR", "FOO_BAR_VALUE"));

            matchEntry = new StorageEntry();
            matchEntry.addAttribute(
                new StorageAttribute(ID, appUid));

            iSession.update(tableName, updateEntry, matchEntry);
            assertTrue("No exp. when non-existent attribute", false);
        }
        catch (StorageException se)
        {
            // PASSED
        }
        catch (Throwable t)
        {
            assertTrue("UnExpected exp6: " + t.toString(), false);
        }

        iSession.close();

        // 7. Test update no connection throws SE.
        try
        {
            tableName = APPLICATION_PACKAGE_TABLE;
            appUid = "UpdateMe";

            updateEntry = new StorageEntry();
            updateEntry.addAttribute(
                new StorageAttribute(VENDOR,
                                     "TestUpdateVal"));

            matchEntry = new StorageEntry();
            matchEntry.addAttribute(
                new StorageAttribute(ID, appUid));

            iSession.update(tableName, updateEntry, matchEntry);
            assertTrue("No exp. when no connection", false);
        }
        catch (StorageException se)
        {
            // PASSED
        }
        catch (Throwable t)
        {
            assertTrue("UnExpected exp7: " + t.toString(), false);
        }
    }

    /**
     * Test update.
     * 1. Test update.
     * 2. Test update whole table.
     * 3. Test update string to INT entry. SE must be thrown.
     * 4. Test update two same entries. Second one is updated.
     * 5. Test update empty value. Entries are updated to storage.
     * 6. Test update null value. Entries are updated to storage.
     * 7. Test update no data in DB. Nothing updated.
     * 8. Test partial update. One attribute has value in DB and other
     *    one does not. Both are updated.
     * 9. Test update primary key.
     */
    public void testUpdate()
    {
        String tableName = APPLICATION_PACKAGE_TABLE;
        StorageEntry updateEntry = null;
        StorageEntry matchEntry = null;
        String appUid = null;

        // 1. Test update.
        try
        {
            iSession.open();
            iSession.startTransaction();
            appUid = "JavaTestUpdate1";

            assertTrue("Populate", iJtu.populateWithDefault(
                           iSession, appUid));

            assertTrue("Check population", iJtu.checkDefaultParams(
                           iSession, appUid));

            updateEntry = new StorageEntry();
            updateEntry.addAttribute(
                new StorageAttribute(VENDOR,
                                     "UpdatedSuiteVenor"));

            // This is left empty as test utils add appUid as match attribute.
            matchEntry = new StorageEntry();

            assertTrue("Update", iJtu.update(iSession,
                                             tableName,
                                             appUid,
                                             updateEntry,
                                             matchEntry));

            assertTrue("Check failed", iJtu.checkParams(iSession,
                       tableName,
                       updateEntry,
                       1));
            iSession.rollbackTransaction();
        }
        catch (Throwable t)
        {
            iSession.rollbackTransaction();
            assertTrue("UnExpected exp1: " + t.toString(), false);
        }

        iSession.close();

        // 2. Test update whole table.
        try
        {
            iSession.open();
            iSession.startTransaction();
            appUid = "JavaTestUpdate2";

            StorageEntry entry = iJtu.createAppPackageEntry();

            assertTrue("Populate", iJtu.populate(
                           iSession, tableName, appUid, entry));

            assertTrue("Check population", iJtu.checkParams(
                           iSession, tableName, entry, 1));

            updateEntry = new StorageEntry();

            updateEntry.addAttribute(new StorageAttribute(
                                         PACKAGE_NAME, "UDPMIDlet"));
            updateEntry.addAttribute(new StorageAttribute(
                                         VENDOR, "UDPVendor"));
            updateEntry.addAttribute(new StorageAttribute(
                                         VERSION, "9.9.9.9"));
            updateEntry.addAttribute(new StorageAttribute(
                                         ROOT_PATH, "/UDPAPI/path"));
            updateEntry.addAttribute(new StorageAttribute(
                                         MEDIA_ID,
                                         "899089",
                                         StorageAttribute.INT_TYPE));
            updateEntry.addAttribute(new StorageAttribute(
                                         INITIAL_SIZE,
                                         "65464",
                                         StorageAttribute.INT_TYPE));
            updateEntry.addAttribute(new StorageAttribute(
                                         JAD_PATH, "UDPAPI/JAD/"));
            updateEntry.addAttribute(new StorageAttribute(
                                         JAR_PATH, "UDPAPI/JAR/"));
            updateEntry.addAttribute(new StorageAttribute(
                                         ACCESS_POINT, "Streaming AP"));

            // This is left empty as test utils add appUid as match attribute.
            matchEntry = new StorageEntry();

            assertTrue("Update", iJtu.update(iSession,
                                             tableName,
                                             appUid,
                                             updateEntry,
                                             matchEntry));

            assertTrue("Check failed", iJtu.checkParams(iSession,
                       tableName,
                       updateEntry,
                       1));
            iSession.rollbackTransaction();
        }
        catch (Throwable t)
        {
            iSession.rollbackTransaction();
            assertTrue("UnExpected exp2: " + t.toString(), false);
        }

        iSession.close();

        // 3. Test update string to INT entry. SE must be thrown.
        try
        {
            iSession.open();
            iSession.startTransaction();
            appUid = "JavaTestUpdate3";

            StorageEntry entry = new StorageEntry();

            entry.addAttribute(new StorageAttribute(
                                   ID, appUid));

            entry.addAttribute(new StorageAttribute(
                                   PACKAGE_NAME, "INT_MIDlet"));

            entry.addAttribute(new StorageAttribute(
                                   INITIAL_SIZE, "12"));

            assertTrue("Populate", iJtu.populate(
                           iSession, tableName, appUid, entry));

            assertTrue("Check population", iJtu.checkParams(
                           iSession, tableName, entry, 1));

            updateEntry = new StorageEntry();
            updateEntry.addAttribute(
                new StorageAttribute(INITIAL_SIZE,
                                     "MyGreatStringID",
                                     StorageAttribute.INT_TYPE));

            iSession.update(tableName, updateEntry, entry);
            assertTrue("No exp update String to INT", false);
        }
        catch (StorageException se)
        {
            // PASSED
        }
        catch (Throwable t)
        {
            iSession.rollbackTransaction();
            assertTrue("UnExpected exp3: " + t.getMessage(), false);
        }

        iSession.close();

        // 4. Test update two same entries. Last one is updated.
        try
        {
            iSession.open();
            iSession.startTransaction();
            appUid = "JavaTestUpdate4";

            StorageEntry entry = new StorageEntry();

            entry.addAttribute(new StorageAttribute(
                                   ID, appUid));

            entry.addAttribute(new StorageAttribute(
                                   PACKAGE_NAME, "INT_MIDlet"));

            entry.addAttribute(new StorageAttribute(
                                   INITIAL_SIZE, "12"));

            assertTrue("Populate", iJtu.populate(
                           iSession, tableName, appUid, entry));

            assertTrue("Check population", iJtu.checkParams(
                           iSession, tableName, entry, 1));

            updateEntry = new StorageEntry();
            updateEntry.addAttribute(
                new StorageAttribute(INITIAL_SIZE,
                                     "1"));
            updateEntry.addAttribute(
                new StorageAttribute(INITIAL_SIZE,
                                     "2"));

            // This is left empty as test utils add appUid as match attribute.
            matchEntry = new StorageEntry();

            assertTrue("Update", iJtu.update(iSession,
                                             tableName,
                                             appUid,
                                             updateEntry,
                                             matchEntry));


            // Remove the first same entry as last one should be updated.
            updateEntry.removeAttribute(
                new StorageAttribute(INITIAL_SIZE, "1"));

            assertTrue("Check failed2", iJtu.checkParams(iSession,
                       tableName,
                       updateEntry,
                       1));
            iSession.rollbackTransaction();
        }
        catch (Throwable t)
        {
            iSession.rollbackTransaction();
            assertTrue("UnExpected exp4: " + t.toString(), false);
        }

        iSession.close();

        // 5. Test update empty value. Entries are updated to storage.
        try
        {
            iSession.open();
            iSession.startTransaction();
            appUid = "JavaTestUpdate1";

            assertTrue("Populate", iJtu.populateWithDefault(
                           iSession, appUid));

            assertTrue("Check population", iJtu.checkDefaultParams(
                           iSession, appUid));

            updateEntry = new StorageEntry();
            updateEntry.addAttribute(
                new StorageAttribute(VENDOR,
                                     ""));

            // This is left empty as test utils add appUid as match attribute.
            matchEntry = new StorageEntry();

            assertTrue("Update", iJtu.update(iSession,
                                             tableName,
                                             appUid,
                                             updateEntry,
                                             matchEntry));

            // Check only updated column
            assertTrue("Check failed", iJtu.checkParams(iSession,
                       tableName,
                       updateEntry,
                       1));

            matchEntry = new StorageEntry();
            matchEntry.addAttribute(new StorageAttribute(ID, appUid));

            // Check other params are intact.
            assertTrue("Check failed", iJtu.checkParams(iSession,
                       tableName,
                       matchEntry,
                       1));

            iSession.rollbackTransaction();
        }
        catch (Throwable t)
        {
            iSession.rollbackTransaction();
            assertTrue("UnExpected exp5: " + t.toString(), false);
        }

        iSession.close();

        // 6. Test update null value. Entries are updated to storage.
        try
        {
            iSession.open();
            iSession.startTransaction();
            appUid = "JavaTestUpdate6";

            assertTrue("Populate", iJtu.populateWithDefault(
                           iSession, appUid));

            assertTrue("Check population", iJtu.checkDefaultParams(
                           iSession, appUid));

            updateEntry = new StorageEntry();
            String value = null;
            updateEntry.addAttribute(
                new StorageAttribute(VENDOR,
                                     value,
                                     StorageAttribute.NULL_TYPE));

            // This is left empty as test utils add appUid as match attribute.
            matchEntry = new StorageEntry();

            assertTrue("Update", iJtu.update(iSession,
                                             tableName,
                                             appUid,
                                             updateEntry,
                                             matchEntry));

            assertTrue("Check failed", iJtu.checkParams(iSession,
                       tableName,
                       updateEntry,
                       1) == false);

            matchEntry = new StorageEntry();
            matchEntry.addAttribute(new StorageAttribute(ID, appUid));

            // Check other params are intact.
            assertTrue("Check failed", iJtu.checkParams(iSession,
                       tableName,
                       matchEntry,
                       1));
            iSession.rollbackTransaction();
        }
        catch (Throwable t)
        {
            iSession.rollbackTransaction();
            assertTrue("UnExpected exp6: " + t.toString(), false);
        }

        iSession.close();

        // 7. Test update no data in DB. Nothing updated.
        try
        {
            iSession.open();
            iSession.startTransaction();
            appUid = "TestUpdate7";

            updateEntry = new StorageEntry();
            String value = null;
            updateEntry.addAttribute(
                new StorageAttribute(VENDOR,
                                     value,
                                     StorageAttribute.NULL_TYPE));

            matchEntry = new StorageEntry();

            matchEntry.addAttribute(
                new StorageAttribute(VENDOR,
                                     value,
                                     StorageAttribute.NULL_TYPE));

            iSession.update(tableName, updateEntry, matchEntry);
            iSession.rollbackTransaction();
        }
        catch (Throwable t)
        {
            iSession.rollbackTransaction();
            assertTrue("UnExpected exp7: " + t.toString(), false);
        }

        iSession.close();

        // 8. Test partial update. One attribute has value in DB and other
        //    one does not. Both are updated.
        try
        {
            iSession.open();
            iSession.startTransaction();
            appUid = "JavaTestUpdate8";

            assertTrue("Populate", iJtu.populateWithDefault(
                           iSession, appUid));

            assertTrue("Check population", iJtu.checkDefaultParams(
                           iSession, appUid));

            updateEntry = new StorageEntry();
            updateEntry.addAttribute(
                new StorageAttribute(VENDOR,
                                     "UpdatedSuiteVenor"));
            // This is not default attribute
            updateEntry.addAttribute(
                new StorageAttribute(JAR_PATH,
                                     "HooooHaaa!!!"));

            // This is left empty as test utils add appUid as match attribute.
            matchEntry = new StorageEntry();

            assertTrue("Update", iJtu.update(iSession,
                                             tableName,
                                             appUid,
                                             updateEntry,
                                             matchEntry));

            assertTrue("Check failed", iJtu.checkParams(iSession,
                       tableName,
                       updateEntry,
                       1));
            iSession.rollbackTransaction();
        }
        catch (Throwable t)
        {
            iSession.rollbackTransaction();
            assertTrue("UnExpected exp8: " + t.toString(), false);
        }

        iSession.close();

        // 9. Test update primary key.
        try
        {
            iSession.open();
            iSession.startTransaction();
            appUid = "JavaTestUpdate9";

            assertTrue("Populate", iJtu.populateWithDefault(
                           iSession, appUid));

            assertTrue("Check population", iJtu.checkDefaultParams(
                           iSession, appUid));

            updateEntry = new StorageEntry();
            updateEntry.addAttribute(
                new StorageAttribute(ID,
                                     "UPDATED_UID"));

            // This is left empty as test utils add appUid as match attribute.
            matchEntry = new StorageEntry();

            assertTrue("Update", iJtu.update(iSession,
                                             tableName,
                                             appUid,
                                             updateEntry,
                                             matchEntry));

            assertTrue("Check failed", iJtu.checkParams(iSession,
                       tableName,
                       updateEntry,
                       1));
            iSession.rollbackTransaction();
        }
        catch (Throwable t)
        {
            iSession.rollbackTransaction();
            assertTrue("UnExpected exp9: " + t.toString(), false);
        }

        iSession.close();
    }
}

File: javacommons/javastorage/tsrc/storageclient/src/testupdate.cpp
/*
* Copyright (c) 2009 Nokia Corporation and/or its subsidiary(-ies).
* All rights reserved.
* This component and the accompanying materials are made available
* under the terms of "Eclipse Public License v1.0"
* which accompanies this distribution, and is available
* at the URL "http://www.eclipse.org/legal/epl-v10.html".
*
* Initial Contributors:
* Nokia Corporation - initial contribution.
*
* Contributors:
*
* Description:
*
*/

#include <stdio.h>
#include <string>

#include "TestHarness.h"

#include "javastorage.h"
#include "javastorageexception.h"
#include "javastoragenames.h"
#include "storagetestutils.h"

using namespace std;
using namespace java::storage;

TEST_GROUP(TestUpdate)
{
    JavaStorage* js;
    JavaStorageTestUtils* jtu;

    TEST_SETUP()
    {
        js = JavaStorage::createInstance();
        jtu = new JavaStorageTestUtils();
    }
    TEST_TEARDOWN()
    {
        try
        {
            js->rollbackTransaction();
            js->close();
        }
        catch (...)
        {
            // No can do
        }

        delete js;
        js = 0;
        delete jtu;
        jtu = 0;
    }
};

/**
 * 1. Test empty tableName throws JSE.
 * 2. Test empty updateEntry throws JSE.
 * 3. Test empty matchEntry throws JSE.
 * 4. Test non-existent tableName throws JSE.
 * 5. Test update non-existent updateEntry name throws JSE.
 * 6. Test update non-existent matchEntry name throws JSE.
 * 7. Test update when connection already closed throws JSE.
 */
TEST(TestUpdate, TestUpdateArguments)
{
    JavaStorageApplicationEntry_t updateEntry;
    JavaStorageApplicationEntry_t matchEntry;
    JavaStorageApplicationEntry_t emptyEntry;

    jtu->createDefaultEntry(updateEntry);
    jtu->createDefaultEntry(matchEntry);

    // 1. Test empty tableName throws JSE.
    string dbName = JAVA_DATABASE_NAME;
    js->open(dbName);

    string empty = "";
    string tableName = APPLICATION_PACKAGE_TABLE;

    try
    {
        js->update(empty, updateEntry, matchEntry);
        FAIL("No exception while update with empty tableName\n");
    }
    catch (JavaStorageException& jse)
    {
        CHECK(EInvalidArgument == jse.mStatus);
    }
    catch (...)
    {
        FAIL("Unexpected exp while update with empty tableName\n");
    }

    // 2. Test empty updateEntry throws JSE.
    try
    {
        js->update(empty, emptyEntry, matchEntry);
        FAIL("No exception while update with empty updateEntry\n");
    }
    catch (JavaStorageException& jse)
    {
        CHECK(EInvalidArgument == jse.mStatus);
    }
    catch (...)
    {
        FAIL("Unexpected exp while update with empty primaryKey\n");
    }

    // 3. Test empty match entry throws JSE.
    try
    {
        js->update(tableName, updateEntry, emptyEntry);
        FAIL("No exception while update with empty updateEntry\n");
    }
    catch (JavaStorageException& jse)
    {
        CHECK(EInvalidArgument == jse.mStatus);
    }
    catch (...)
    {
        FAIL("Unexpected exp while update with empty entry\n");
    }

    // 4. Test non-existent tableName throws JSE.
    try
    {
        string invalidTableName = "FOO_BAAR_TAABLE";
        js->update(invalidTableName, updateEntry, matchEntry);
        FAIL("No exception while update with non-existing tableName\n");
    }
    catch (JavaStorageException& jse)
    {
        // PASS
    }
    catch (...)
    {
        FAIL("Unexpected exp while update with non-existing tableName\n");
    }

    // 5. Test update non-existent updateEntry throws JSE.
    try
    {
        JavaStorageApplicationEntry_t nonExistentEntry;
        JavaStorageEntry entry;
        entry.setEntry(L"FOO_BAAR", L"BAAR_FOO_VALUE");
        nonExistentEntry.insert(entry);

        js->update(tableName, nonExistentEntry, matchEntry);
        FAIL("No exception while update non-existent updateEntry name\n");
    }
    catch (JavaStorageException& jse)
    {
        // PASS
    }
    catch (...)
    {
        FAIL("Unexp exp while update with non-existent updateEntry name\n");
    }

    // 6. Test update non-existent matchEntry throws JSE.
    try
    {
        JavaStorageApplicationEntry_t nonExistentEntry;
        JavaStorageEntry entry;
        entry.setEntry(L"FOO_BAAR", L"BAAR_FOO_VALUE");
        nonExistentEntry.insert(entry);

        js->update(tableName, updateEntry, nonExistentEntry);
        FAIL("No exception while update non-existent matchEntry name\n");
    }
    catch (JavaStorageException& jse)
    {
        // PASS
    }
    catch (...)
    {
        FAIL("Unexpected exp while update with non-existent matchentry name\n");
    }

    // 7. Test update when connection already closed throws JSE.
    js->close();
    try
    {
        js->update(tableName, updateEntry, matchEntry);
        FAIL("No exception while update with closed connection\n");
    }
    catch (JavaStorageException& jse)
    {
        CHECK(EInvalidConnection == jse.mStatus);
    }
    catch (...)
    {
        FAIL("Unexpected exp while update with closed connection\n");
    }
    updateEntry.clear();
    matchEntry.clear();
}

/**
 * 1. Test update.
 * 2. Test update one entry.
 * 3. Test update string to int entry. JSE is thrown.
 * 4. Test update same key twice in same statemenet. Second attribute is updated
 *    to storage.
 * 5. Test update no name. JSE is thrown.
 * 6. Test update no value. Value is cleared.
 * 7. Test update no existing value. Data is written to storage. Storage must
 *    contain some data where matchEntry matches but actual entry in the
 *    statement is not written to storage.
 * 8. Test update no existing entries by primary key. No entries updated.
 * 9. Test update whole table.
 * 10. Test Update primary key.
 * 11. Test update to NULL.
 */
TEST(TestUpdate, TestUpdate)
{
    string tableName = APPLICATION_PACKAGE_TABLE;
    JavaStorageApplicationEntry_t appEntry;
    JavaStorageApplicationEntry_t idEntry;

    JavaStorageEntry entry;
    entry.setEntry(JAD_PATH, L"/usr/local/bin/myApps/");
    appEntry.insert(entry);

    entry.setEntry(JAR_PATH, L"/usr/share/private/[abcs12345]/myApps/");
    appEntry.insert(entry);

    entry.setEntry(INITIAL_SIZE, L"16543", JavaStorageEntry::INT);
    appEntry.insert(entry);

    JavaStorageApplicationEntry_t defaultEntry;
    jtu->createDefaultEntry(defaultEntry);
    entry.setEntry(ID, L"TestUpdate1");
    defaultEntry.insert(entry);
    idEntry.insert(entry);

    try
    {
        js->open();
        js->startTransaction();

        CHECK(jtu->populate(*js, tableName, defaultEntry));
        CHECK(jtu->checkParams(*js, tableName, defaultEntry, 1));

        js->update(tableName, appEntry, defaultEntry);
        CHECK(jtu->checkParams(*js, tableName, appEntry, 1));

        // This comes from default params. This ensures that both values are
        // written under same primary key.
        entry.setEntry(PACKAGE_NAME, L"BasicGame");
        appEntry.insert(entry);

        CHECK(jtu->checkParams(*js, tableName, appEntry, 1));
    }
    catch (...)
    {
        FAIL("Unexpected exp while update\n");
    }

    // 2. Test update one entry.
    try
    {
        appEntry.clear();
        entry.setEntry(PACKAGE_NAME, L"NewName");
        appEntry.insert(entry);

        js->update(tableName, appEntry, defaultEntry);
        CHECK(jtu->checkParams(*js, tableName, appEntry, 1));
    }
    catch (...)
    {
        FAIL("Unexpected exp while update one entry\n");
    }

    // 3. Test update string to int entry. JSE is thrown.
    try
    {
        appEntry.clear();
        entry.setEntry(INITIAL_SIZE, L"BIG_SIZE", JavaStorageEntry::INT);
        appEntry.insert(entry);

        js->update(tableName, appEntry, defaultEntry);
        FAIL("No exp. while updating string to int column\n");
    }
    catch (JavaStorageException& jse)
    {
        // PASSED
    }
    catch (...)
    {
        FAIL("Unexpected exp while update string to int column\n");
    }

    // 4. Test update same key twice in same statemenet. Second attribute
    // is considered. Logically it is updated latests.
    try
    {
        appEntry.clear();
        entry.setEntry(JAD_PATH, L"First path");
        appEntry.insert(entry);

        entry.setEntry(JAD_PATH, L"Second path");
        appEntry.insert(entry);

        // Use idEntry from now on as defaultEntry is modified in the storage.
        js->update(tableName, appEntry, idEntry);

        JavaStorageApplicationEntry_t checkEntry;
        entry.setEntry(JAD_PATH, L"Second path");
        checkEntry.insert(entry);

        CHECK(jtu->checkParams(*js, tableName, checkEntry, 1));
    }
    catch (...)
    {
        FAIL("Unexpected exp while update same entry twice\n");
    }

    // 5. Test update no name. JSE is thrown.
    try
    {
        appEntry.clear();
        entry.setEntry(L"", L"12345");
        appEntry.insert(entry);

        js->update(tableName, appEntry, idEntry);
        FAIL("No exp. while updating no name\n");
    }
    catch (JavaStorageException& jse)
    {
        // PASSED
    }
    catch (...)
    {
        FAIL("Unexpected exp while update no name\n");
    }

    // 6. Test update no value. Value is cleared.
    try
    {
        appEntry.clear();
        entry.setEntry(PACKAGE_NAME, L"");
        appEntry.insert(entry);

        js->update(tableName, appEntry, idEntry);
        CHECK(jtu->checkParams(*js, tableName, idEntry, 1));
        CHECK(jtu->checkParams(*js, tableName, appEntry, 1));
    }
    catch (...)
    {
        FAIL("Unexp exp while update no value\n");
    }

    // 7. Test update no existing value.
    try
    {
        appEntry.clear();
        entry.setEntry(ROOT_PATH, L"NonExistingValue");
        appEntry.insert(entry);

        js->update(tableName, appEntry, idEntry);
        CHECK(jtu->checkParams(*js, tableName, appEntry, 1));
    }
    catch (...)
    {
        FAIL("Unexpected exp while update no value\n");
    }

    // 8. Test update no matching entries
    try
    {
        appEntry.clear();
        entry.setEntry(ROOT_PATH, L"UNIQUE_ID_X");
        appEntry.insert(entry);

        js->update(tableName, defaultEntry, appEntry);

        // Verify not updated.
        CHECK(jtu->checkParams(*js, tableName, defaultEntry, 0));
    }
    catch (...)
    {
        FAIL("Unexpected exp while update with non-existent primaryKey\n");
    }

    // 9. Test update whole table.
    try
    {
        appEntry.clear();
        // Add also ID createAppDataEntry does not create it
        appEntry = idEntry;
        jtu->createAppDataEntry(appEntry);

        js->update(tableName, appEntry, idEntry);
        CHECK(jtu->checkParams(*js, tableName, appEntry, 1));
    }
    catch (...)
    {
        FAIL("Unexpected exp while update whole table\n");
    }

    // 10. Test Update primary key.
    try
    {
        appEntry.clear();
        // ID act as primary key on this table.
        entry.setEntry(ID, L"TestUpdate2");
        appEntry.insert(entry);

        js->update(tableName, appEntry, idEntry);
        CHECK(jtu->checkParams(*js, tableName, appEntry, 1));

        // Check it was really update that old ID is not found.
        CHECK(jtu->checkParams(*js, tableName, idEntry, 0));
    }
    catch (...)
    {
        FAIL("Unexpected exp while update primary key\n");
    }

    // 11. Test update to NULL.
    try
    {

        appEntry.clear();
        idEntry.clear();

        entry.setEntry(ID, L"TestUpdate2");
        idEntry.insert(entry);

        entry.setEntry(PACKAGE_NAME, L"", JavaStorageEntry::NULL_TYPE);
        appEntry.insert(entry);

        entry.setEntry(JAD_URL, L"FindMe!");
        appEntry.insert(entry);

        js->update(tableName, appEntry, idEntry);
        CHECK(jtu->checkParams(*js, tableName, idEntry, 1));
        // Set to NULL.
        CHECK(jtu->checkParams(*js, tableName, appEntry, 1) == false);

        /*appEntry.clear();
        // ID act as primary key on this table.
        entry.setEntry(ID, L"TestUpdate2");
        appEntry.insert(entry);

        js->update(tableName, appEntry, idEntry);
        CHECK(jtu->checkParams(*js, tableName, appEntry, 1));

        // Check it was really update that old ID is not found.
        CHECK(jtu->checkParams(*js, tableName, idEntry, 0));*/
    }
    catch (...)
    {
        FAIL("Unexpected exp while update primary key\n");
    }


    appEntry.clear();
    idEntry.clear();
    defaultEntry.clear();
    js->rollbackTransaction();
    js->close();
}
